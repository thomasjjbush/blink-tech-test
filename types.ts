import { AnyAction } from 'redux';
import { ThunkAction, ThunkDispatch } from 'redux-thunk';

export type Action = ThunkAction<Promise<AnyAction | void>, Store, unknown, AnyAction>;

export type Dispatch = ThunkDispatch<Store, void, AnyAction>;

export interface Conversation {
    id: string;
    last_updated: string;
    name: string;
}

export enum ConversationActions {
    LOADED = 'conversation/LOADED',
    UPDATED = 'conversation/UPDATED',
}

export interface ConversationsProps {
    conversation: string;
    setConversation: (conversation: string) => void;
}

export enum Endpoints {
    CONVERSATIONS = 'conversations',
    MESSAGES = 'messages',
}

export interface ErrorState {
    code?: number;
    fatal?: boolean;
    message?: string;
}

export enum ErrorActions {
    'FATAL' = 'error/FATAL',
}

export interface Message {
    conversation: string;
    edited?: true;
    id: string;
    isUser?: true;
    last_updated: string;
    text: string;
}

export enum MessageActions {
    EDIT = 'messages/EDIT',
    LOADED = 'messages/LOADED',
    SEND = 'messages/SEND',
}

export interface MessageBubbleProps extends Message {
    onClick: (id: string, text: string) => void;
}

export interface MessageState {
    [key: string]: Message[];
}

export interface MessagesProps {
    conversation: string;
}

export interface Options {
    [key: string]: string | number;
    conversation?: string;
    _sort?: string;
}

export interface Store {
    conversations: Conversation[];
    error: ErrorState;
    messages: MessageState;
}

export type StyledProps<Props = unknown> = { theme: Theme } & Props;

export interface TextAreaProps {
    setValue: (value: string) => void;
    value: string;
}

export interface Theme {
    brightBlue: string;
    darkBlue: string;
    lightBlue: string;
}
