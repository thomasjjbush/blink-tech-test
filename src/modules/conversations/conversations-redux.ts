import { getData } from '../../utils/get-data/get-data';
import { Action, Conversation, ConversationActions, Endpoints, ErrorActions, MessageActions } from '../../../types';
import { AnyAction } from 'redux';

export const loadConversations = (): Action => async (dispatch) => {
    try {
        const payload = await getData<Conversation[]>(Endpoints.CONVERSATIONS);
        dispatch({ type: ConversationActions.LOADED, payload });
    } catch (err) {
        dispatch({ type: ErrorActions.FATAL, payload: err });
    }
};

export const conversationsReducer = (state: Conversation[] = [], { payload, type }: AnyAction): Conversation[] => {
    switch (type) {
        case ConversationActions.LOADED:
            return payload;
        case MessageActions.SEND:
            return state.map((convo) =>
                convo.id === payload.conversation ? { ...convo, last_updated: payload.last_updated } : convo,
            );
        default:
            return state;
    }
};
