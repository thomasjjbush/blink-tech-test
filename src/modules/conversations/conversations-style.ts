import styled from 'styled-components';
import { StyledProps } from '../../../types';

export const Conversation = styled.button<StyledProps<{ active: boolean }>>`
    background-color: ${({ active, theme }) => (active ? theme.brightBlue : 'transparent')};
    border: 0;
    color: white;
    cursor: pointer;
    display: block;
    padding: 20px;
    text-align: left;
    width: 100%;

    &:not(:last-of-type) {
        border-bottom: solid 2px white;
    }
`;

export const Conversations = styled.div<StyledProps>`
    background-color: ${({ theme }) => theme.darkBlue};
`;

export const Name = styled.h2`
    margin: 0 0 10px;
`;
