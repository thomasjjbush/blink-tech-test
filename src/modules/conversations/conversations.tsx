import React, { FC, ReactElement, useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { loadConversations } from './conversations-redux';
import { DateAndTime } from '../../components/date-and-time/date-and-time';
import { Conversation, ConversationsProps as Props, Store } from '../../../types';
import * as Styled from './conversations-style';

export const Conversations: FC<Props> = ({ conversation, setConversation }: Props): ReactElement => {
    const dispatch = useDispatch();
    const conversations = useSelector<Store, Conversation[]>((state) => state.conversations);

    useEffect(() => {
        (async () => {
            await dispatch(loadConversations());
        })();
    }, []);

    return (
        <Styled.Conversations>
            {conversations
                .sort((a, b) => new Date(b.last_updated).valueOf() - new Date(a.last_updated).valueOf())
                .map(({ id, name, last_updated }) => (
                    <Styled.Conversation
                        active={conversation === id}
                        key={id}
                        onClick={() => setConversation(id)}
                        type="button"
                    >
                        <Styled.Name>{name}</Styled.Name>
                        <DateAndTime>{last_updated}</DateAndTime>
                    </Styled.Conversation>
                ))}
        </Styled.Conversations>
    );
};
