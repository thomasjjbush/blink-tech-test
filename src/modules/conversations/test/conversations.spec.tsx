import React from 'react';
import { mount, shallow } from 'enzyme';
import { Conversations } from './../conversations';
import { loadConversations } from './../conversations-redux';
import { ConversationsProps as Props } from '../../../../types';

const conversations = [
    {
        id: '423',
        last_updated: new Date('Jan 1, 1970'),
        name: 'peter griffin',
    },
    {
        id: '5453',
        last_updated: new Date('Jan 2, 1970'),
        name: 'grandma',
    },
];

jest.mock('react-redux', () => ({
    useDispatch: jest.fn().mockReturnValue(jest.fn()),
    useSelector: jest.fn((cb) => cb({ conversations })),
}));

jest.mock('./../conversations-redux', () => ({ loadConversations: jest.fn().mockReturnValue('loadConversations') }));

describe('Conversation', () => {
    let props: Props;

    beforeEach(() => {
        jest.clearAllMocks();
        props = {
            conversation: '423',
            setConversation: jest.fn(),
        };
    });

    it('should match snapshot and apply correct conditional styles to conversation button', () => {
        expect(shallow(<Conversations {...props} />)).toMatchSnapshot();
    });

    it('should sort the items into the correct order of oldest first', () => {
        const wrapper = shallow(<Conversations {...props} />);
        const names = wrapper.find({ 'data-test-id': 'conversation-name' });

        expect(names.first().text()).toBe('grandma');
        expect(names.last().text()).toBe('peter griffin');
    });

    it('should invoke setConversation prop on button click with correct id', () => {
        const wrapper = shallow(<Conversations {...props} />);
        const buttons = wrapper.find({ type: 'button' });

        buttons.first().simulate('click');

        expect(props.setConversation).toHaveBeenCalledTimes(1);
        expect(props.setConversation).toHaveBeenCalledWith('5453');
    });

    it('should loadConversations on mount', async () => {
        mount(<Conversations {...props} />);

        expect(loadConversations).toHaveBeenCalledTimes(1);
        expect(loadConversations).toHaveBeenCalledWith();
    });
});
