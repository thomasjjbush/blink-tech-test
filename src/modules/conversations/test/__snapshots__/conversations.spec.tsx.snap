// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Conversation should match snapshot and apply correct conditional styles to conversation button 1`] = `
<styled.div
  styles="
    background-color: theme.darkBlue;
"
>
  <styled.button
    active={false}
    key="5453"
    onClick={[Function]}
    styles="
    background-color: transparent;
    border: 0;
    color: white;
    cursor: pointer;
    display: block;
    padding: 20px;
    text-align: left;
    width: 100%;

    &:not(:last-of-type) {
        border-bottom: solid 2px white;
    }
"
    type="button"
  >
    <styled.h2
      data-test-id="conversation-name"
      styles="
    margin: 0 0 10px;
"
    >
      grandma
    </styled.h2>
    <DateAndTime>
      <Component />
    </DateAndTime>
  </styled.button>
  <styled.button
    active={true}
    key="423"
    onClick={[Function]}
    styles="
    background-color: theme.brightBlue;
    border: 0;
    color: white;
    cursor: pointer;
    display: block;
    padding: 20px;
    text-align: left;
    width: 100%;

    &:not(:last-of-type) {
        border-bottom: solid 2px white;
    }
"
    type="button"
  >
    <styled.h2
      data-test-id="conversation-name"
      styles="
    margin: 0 0 10px;
"
    >
      peter griffin
    </styled.h2>
    <DateAndTime>
      <Component />
    </DateAndTime>
  </styled.button>
</styled.div>
`;
